Bug/dirty telemetry attributes

nickperry12 opened the pull request:
## Description
- Added a function that intercepts the `onFinish` callback to capture telemetry data when streams complete
- Streams return immediately while telemetry is captured asynchronously on completion
- This fix works for both `agent.stream()` and `agent.streamVNext()`
- Existing behaviours are preserved for backwards compatibility

## Related Issue(s)

Fixes #5870

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [ ] Test update

## Checklist

- [ ] I have made corresponding changes to the documentation (if applicable)
- [ ] I have added tests that prove my fix is effective or that my feature works

nickperry12 committed:
Fixed telemetry attributes as promise status description when `agent.stream` is called

- Fixed the core issue of telemetry serializing promise status when calls to `agent.stream` were made
- Added a function that intercepts `onFinish` callbacks to capture telemetry data when streams complete
- Added `StreamFinishData`, `StreamOptions` and `EnhancedSpan` interfaces
- Preserves existing behaviors for backwards compatibility

nickperry12 committed:
Added changeset

nickperry12 commented on packages/core/src/telemetry/telemetry.decorators.ts:
```diff
@@ -56,51 +145,63 @@ export function withSpan(options: {
+      let result: unknown;
       try {
+        // For streaming methods, enhance arguments with telemetry capture before calling
+        const enhancedArgs = isStreamingResult(result, methodName)
```
@greptileai I think there's a misunderstanding. The method won't return false here, as the first check is for the method name, and returns true immediately if there are method name matches for known streaming methods

DanielSLew committed:
Merge branch 'main' into bug/dirty-telemetry-attributes

DanielSLew committed:
Merge branch 'main' into bug/dirty-telemetry-attributes

abhiaiyer91 committed:
Merge branch 'main' into bug/dirty-telemetry-attributes

nickperry12 committed:
Fixed linting issues

nickperry12 committed:
Merge branch 'bug/dirty-telemetry-attributes' of https://github.com/nickperry12/mastra into bug/dirty-telemetry-attributes

DanielSLew committed:
Merge branch 'main' into bug/dirty-telemetry-attributes

wardpeet reviewed (commented):
The actual implementation looks good but what I would do to improve the readability of this PR is to add `stream` to the exclude list inside the agent class.

```
@InstrumentClass({
  prefix: 'agent',
  excludeMethods: [
    'hasOwnMemory',
    'getMemory',
    '__primitive',
    '__registerMastra',
    '__registerPrimitives',
    '__setTools',
    '__setLogger',
    '__setTelemetry',
    'stream',
    'log',
    'getModel',
    'getInstructions',
    'getTools',
    'getLLM',
    'getWorkflows',
    'getDefaultGenerateOptions',
    'getDefaultStreamOptions',
    'getDescription',
  ],
})
```

Inside the stream i would manually call the withSpan  or duplicate the logic so it's nicely inside the caller function. This improves readabilty but also awareness that we should check any stream call that the telemetry is updated correctly

wardpeet commented:
@nickperry12 yes, seems like it

wardpeet committed:
fix success message

DanielSLew committed:
Update packages/core/src/telemetry/telemetry.ts

DanielSLew committed:
Update packages/core/src/telemetry/telemetry.decorators.ts

wardpeet commented:
@DanielSLew this change https://github.com/mastra-ai/mastra/pull/6265#issuecomment-3133492550 should still happen

nickperry12 committed:
refactor:

- manually calling `withSpan` inside of the `stream` method for improved readability
- added `stream` method to the excluded list inside the agent class

Co-Authored-By: Jason Jones <jasonjabarjones@gmail.com>

nickperry12 committed:
reverted changes

Reverted back to previous implementation

Co-Authored-By: Jason Jones <jasonjabarjones@gmail.com>

nickperry12 committed:
Merge branch 'main' into bug/dirty-telemetry-attributes

DanielSLew committed:
Merge branch 'main' into bug/dirty-telemetry-attributes
